File: frontend/src/lib/order/Order.ts
import { z } from 'zod'

export const OrderProductSchema = z.object({
  id: z.number().int().min(1),
  name: z.string().min(1).max(100),
  netPriceCents: z.number().int().min(0),
  quantity: z.number().int().min(1),
})
export type OrderProduct = z.infer<typeof OrderProductSchema>

export const PlaceOrderSchema = z.object({
  tableId: z.number().int().min(1),
  products: OrderProductSchema.array().min(1),
})
export type PlaceOrder = z.infer<typeof PlaceOrderSchema>

export const OrderSchema = z.object({
  id: z.uuid(),
  userId: z.number().int().min(1),
  tableId: z.number().int().min(1),
  products: OrderProductSchema.array().min(1),
  totalNetPriceCents: z.number().int().min(0),
  placedAt: z.string().refine((date) => !isNaN(Date.parse(date)), {
    message: 'Invalid date format',
  }),
})
export type Order = z.infer<typeof OrderSchema>

File: backend/table_service/order.go
package table_service

import (
	"time"

	z "github.com/Oudwins/zog"
	"github.com/nicograef/jotti/backend/product_admin"
)

type orderProduct struct {
	ID            int    `json:"id"`
	Name          string `json:"name"`
	NetPriceCents int    `json:"netPriceCents"`
	Quantity      int    `json:"quantity"`
}

var orderProductSchema = z.Struct(z.Shape{
	"ID":            product_admin.IDSchema.Required(),
	"Name":          product_admin.NameSchema.Required(),
	"NetPriceCents": product_admin.NetPriceCentsSchema.Required(),
	"Quantity":      z.Int().GTE(1, z.Message("Quantity must be at least 1")).Required(),
})

type Order struct {
	ID                 string         `json:"id"`
	UserID             int            `json:"userId"`
	TableID            int            `json:"tableId"`
	Products           []orderProduct `json:"products"`
	TotalNetPriceCents int            `json:"totalNetPriceCents"`
	PlacedAt           time.Time      `json:"placedAt"`
}

var orderSchema = z.Struct(z.Shape{
	"ID":                 z.String().UUID().Required(),
	"UserID":             z.Int().GTE(1).Required(),
	"TableID":            z.Int().GTE(1).Required(),
	"Products":           z.Slice(orderProductSchema).Min(1).Required(),
	"TotalNetPriceCents": z.Int().GTE(0).Required(),
	"PlacedAt":           z.Time().Required(),
})
